### PR TITLE
feat: added support for index include

### LINKED ISSUE #15124: SQL Server - Create Index donot have support for "Include"
Below is the basic command for creating index in SQL Server. The below command also has "INCLUDE" key which sequelize donot support while creating indexing

/*
USE [cube-qa]
GO
CREATE NONCLUSTERED INDEX [<Name of Missing Index, sysname,>]
ON [dbo].[has_roles] ([resourceType])
INCLUDE ([resourceType],[resourceId])
GO
*/

### PR DESCRIPTION
**Pull request by lohart13**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [x] Have you added new tests to prevent regressions?
- [x] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

This pull requests adds the ability to support indexes with included columns for MSSQL.
For information on how this works and the benefits, please see [Microsoft documentation](https://docs.microsoft.com/en-us/sql/relational-databases/indexes/create-indexes-with-included-columns?view=sql-server-ver15) of this feature.

Fixes #15124

### CODE REVIEW

File: src/dialects/mssql/index.js
Comment (WikiRik) [inline src/dialects/mssql/index.js:40]: You should add the include option in the abstract index as well;
https://github.com/sequelize/sequelize/blob/fe3f470194e41677c25504910cb5458a70382529/src/dialects/abstract/index.ts#L82-L92
and 
https://github.com/sequelize/sequelize/blob/fe3f470194e41677c25504910cb5458a70382529/src/dialects/abstract/index.ts#L195-L205

**Review by WikiRik — COMMENTED** [review]
Will need to do a more in-depth review later, but these are my initial comments

File: test/unit/sql/index.test.js
Comment (lohart13) [reply test/unit/sql/index.test.js]: Updated now

File: test/integration/dialects/mssql/errors.test.js
Comment (lohart13) [reply test/integration/dialects/mssql/errors.test.js]: I've changed this to use the `dialect.supports` object

```js
    if (!current.dialect.supports.index.include) {
      return;
    }
```

File: src/dialects/mssql/index.js
Comment (lohart13) [reply src/dialects/mssql/index.js:40]: Added now

**Comment by lohart13** [comment]
@WikiRik looking at the test, do you want this to exist under the dialects > mssql folder?
It appears that db2 and postgres can support index with include (in specific versions) therefore not exactly mssql specific.

- https://www.postgresql.org/docs/current/sql-createindex.html
- https://www.ibm.com/docs/en/db2/11.1?topic=statements-create-index

**Comment by WikiRik** [comment]
I guess you can include the tests after the following;
https://github.com/sequelize/sequelize/blob/79a53009e86df7efe12642aaffaa36a14bd8abc7/test/integration/model.test.js#L612

But I might prefer a separate test file `test/integration/index.test.js` since model.test.js already has quite some tests

**Comment by lohart13** [comment]
> But I might prefer a separate test file `test/integration/index.test.js` since model.test.js already has quite some tests

I've updated the test location now.
Let me know if you have any further questions

**Comment by WikiRik** [comment]
Can you enable the support for db2 and postgres in this PR as well?

**Comment by lohart13** [comment]
@WikiRik I've added support now, however there are some conditions for postgres and db2.

For postgres, the include option for indexes is only supported in version 11 and above therefore this would not be supported for all the versions of postgres that sequelize currently support.

For db2, the include option for indexes is only available for unique indexes.

I've added in some errors and tests to make sure that these limitations are enforced.

Let me know if you have any questions/suggestions

**Comment by WikiRik** [comment]
Thanks! I think we're dropping support for postgres 10 in v7 as well since the final release of that will come out on November 10th so let's wait with merging this PR until we do so. That makes the implementation easier.

Pinging @ephys @sdepold as well to confirm dropping postgres 10 support as soon as postgres itself does it

**Comment by ephys** [comment]
We can drop postgres 10 in v7 
image

**Comment by WikiRik** [comment]
I've created a PR to drop the postgres 10 support; it's #15289. Once that has been merged, this PR can be updated accordingly

**Comment by lohart13** [comment]
I've removed the postgres db version checks now.
LMK if there are any further changes required

File: test/unit/sql/index.test.js
Comment (ephys) [inline test/unit/sql/index.test.js:237]: this test should run on all dialects, but dialects that don't support it should check that an error is thrown. You can do it like this:

```ts
// arrow function so expectsql can catch the error
expectsql(() => sql.addIndexQuery('User',  {
  name: 'email_include_name',
  fields: ['email'],
  include: ['first_name', 'last_name'],
  unique: true,
}), {
  'mysql mariadb': new Error('Your error message here')
  postgres: 'CREATE UNIQUE INDEX [email_include_name] ON [User] ([email]) INCLUDE ([first_name], [last_name])',
});
```

File: test/unit/sql/index.test.js
Comment (ephys) [inline test/unit/sql/index.test.js:242]: could you add a test where `include` is a literal, and `include` is an array of literals? We try to support literal everywhere as an escape hatch

```ts
include: [literal('first_name')],

//

include: literal('(first_name, last_name)'),
```

it should add them as-is without further processing

**Review by ephys — CHANGES_REQUESTED** [review]

File: src/dialects/abstract/query-generator.js
Comment (lohart13) [reply src/dialects/abstract/query-generator.js]: Updated now and added support for literals

File: test/integration/index.test.js
Comment (lohart13) [reply test/integration/index.test.js]: Done

File: test/unit/sql/index.test.js
Comment (lohart13) [reply test/unit/sql/index.test.js]: Done

File: test/unit/sql/index.test.js
Comment (lohart13) [reply test/unit/sql/index.test.js]: Done

**Review by ephys — DISMISSED** [review]
This looks like really good work! Thank you :)

File: test/unit/sql/index.test.js
Comment (WikiRik) [inline test/unit/sql/index.test.js:256]:
```suggestion
```

File: test/unit/sql/index.test.js
Comment (WikiRik) [inline test/unit/sql/index.test.js:270]:
```suggestion
```

File: test/unit/sql/index.test.js
Comment (WikiRik) [inline test/unit/sql/index.test.js:284]:
```suggestion
```

File: test/integration/index.test.ts
Comment (WikiRik) [inline test/integration/index.test.ts:24]:
```suggestion
```

**Review by WikiRik — APPROVED** [review]
These comments are redundant since the error is clear. But it's a non-blocking issue.
Thanks for working on this!

**Review by WikiRik — APPROVED** [review]

**Comment by krzysztofw-hippo** [comment]
Hi, any idea when this will be released? I don't see it in v6.26.0, will it be included in v6.26.1?
Many thanks

**Comment by WikiRik** [comment]
This will land in the next v7 alpha, there are no plans from the maintainers to backport this to v6 but we're always open to PRs from the community

**Comment by lohart13** [comment]
Tagging PR to https://github.com/sequelize/website/issues/501

**Pull request merged**
